core/src/main/java/org/infinispan/stats/impl/CacheContainerStatsImpl.java

```diff
@@ -683,4 +676,18 @@ public Json toJson() {
             .set("average_remove_time_nanos", getAverageRemoveTimeNanos())
             .set("required_minimum_number_of_nodes", getRequiredMinimumNumberOfNodes());
    }
+
+   private final class StatsHolder {
+      final long expiration;
+      final List<Stats> stats;
+
+      StatsHolder(List<Stats> stats) {
+         this.expiration = timeService.expectedEndTime(1, TimeUnit.SECONDS);
```

Comment:
Do Prometheus or other monitoring tools really fetch statistics multiple times per second?
If yes, can't we lower the fetch frequency?
If no, would this change help at all?

Reply:
The problem isn't prometheus it is this https://issues.redhat.com/browse/ISPN-12607?focusedCommentId=15647901&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-15647901

Reply:
> Do Prometheus or other monitoring tools really fetch statistics multiple times per second?

Nope it's usually in the tens of seconds, in our docs we recommend `30s`.

> If yes, can't we lower the fetch frequency?

Sure we can. However, I went with 1 second as I figured that one second wouldn't cause any issues for users who want relatively upto date stats or are using jmx directly.

The key here is that we're preventing each call to the mbean attribute from the {{MetricsCollector}} creating a new {{StatsImpl}} and calling size.

The expectation is that a new StatsHolder will be created per prometheus call to `/metrics`.

Reply:
@ryanemerson I didn't mean increasing the lifespan of the cached stats in your code: I thought we actually had Prometheus fetching our metrics twice a second. Since you say it's actually 30s, my "if yes" question doesn't help.

@wburns thanks for the link, but I'm not convinced the problem is in how `MetricsCollector` registers metrics.
IMO the problem is that a method like `CacheContainerStatsImpl.calculateAverageReadTime()` asks for all the statistics of all the caches, when it just needs `CacheMgmtInterceptor.getAverageReadTime()`.

Reply:
Having said that, caching might make more sense if we did it at the cache level. That way a cache's size wouldn't have to be computed twice, once for the cache's currentNumberOfEntries and then again for the container's currentNumberOfEntries.

Reply:
> @wburns thanks for the link, but I'm not convinced the problem is in how `MetricsCollector` registers metrics.
> IMO the problem is that a method like `CacheContainerStatsImpl.calculateAverageReadTime()` asks for all the statistics of all the caches, when it just needs `CacheMgmtInterceptor.getAverageReadTime()`.

I wasn't saying it was how it registers metrics. The problem is definitely that every time you invoke any cache stat it has to retrieve them all. Unfortunately, all of those invoke Cache.getStats which does that. This PR is just a bandaid so this isn't a complete performance killer.

Reply:
> Having said that, caching might make more sense if we did it at the cache level. That way a cache's size wouldn't have to be computed twice, once for the cache's currentNumberOfEntries and then again for the container's currentNumberOfEntries.

Yeah, I had debated that. However the problem is size and data container are two different things and one is segment aware and the other isn't. It becomes a lot more complicated. Where as this fix is very simple and should get things working right away. There are other JIRAs to actually fix the underlying issues.

Reply:
Like Will says, this is very much a bandaid. We need this for `12.0.0.Final` and `11.0.x` asap. We can work on more elegant solutions in `12.1.x` or `13`

Reply:
Sounds fair @wburns @ryanemerson, just make sure you link to those other JIRAs!